### PR TITLE
Fully replace legacy sidebar implementation with slim sidebar

### LINKED ISSUE #7943: CSS Variables are repeated in core.css and sidebar.css files
### Issue Summary

CSS variables are repeated in both core.css and sidebar.css files.

Extra info: I have some wagtail sites with customized admin colors, updating them to work with wagtail 2.16 I found that CSS variables for colors are repeated in both core.css and sidebar.css files

### Evidence

1. Start a new project with `wagtail start myproject`
2. Go to admin home
3. Open inspector in your browser and check styles panel

Styles in core.css file
image

Styles in sidebar.css file
image

### Technical details

* Python version: Python 3.10.1
* Django version: Django 3.2.15
* Wagtail version: Wagtail 2.16
* Browser version: Chrome 98

### PR DESCRIPTION
**Pull request by thibaudcolas**
Addresses #7995. Also fixes #7943, #8134, #7863, and makes #8202 and #8181 redundant.

I also chose to implement #8199 / #8087, because it does seem like a valid use case when described like this. I could also see the benefit of setting a `aria-label` on links so they make sense without context, if someone needs WCAG AAA compliance. I made the PR in two commits with the removal of the legacy sidebar in one and the attributes support separately, so we can have separate CHANGELOG items for both (and a diff).

Following a UI team discussion, I chose to move the sidebar’s stylesheets to be within `core.scss`. This was the source of a few of those bugs, as the sidebar stylesheet was including some of our styles twice. I kept the JS as a separate file though. I have removed as much of the code as I could find. There might be spots in Python I didn’t handle properly – I’m not too familiar with the dual `render_html` and `render_component` we had for some classes.

- [x] Do the tests still pass?[^1]
- [x] Does the code comply with the style guide? 
    - [x] Run `make lint` from the Wagtail root. 
- [x] For Python changes: Have you added tests to cover the new/fixed behaviour?
- [ ] For front-end changes: Did you test on all of Wagtail’s supported environments?[^2]
    - Tested in Chrome only
- ~~[ ] For new features: Has the documentation been updated accordingly?~~

### CODE REVIEW

File: client/scss/components/_grid.legacy.scss [hunk 1]
```diff
@@ -2,6 +2,16 @@
...
+
+  @include media-breakpoint-up(sm) {
+    @include transition(padding-inline-start $menu-transition-duration ease);
+    transform: none;
+    padding-inline-start: $menu-width;
+
+    .sidebar-collapsed & {
+      padding-inline-start: $menu-width-slim;
+    }
+  }
```
Comment (thibaudcolas) [inline client/scss/components/_grid.legacy.scss:14]: This was previously defined in `_main-nav.scss`, code is the same as before.

File: client/scss/components/_header.scss [hunk 2]
```diff
@@ -49,7 +50,7 @@ header {
 
   .search-form .icon {
     @include svg-icon(1.3rem);
-    color: $nav-search-color;
+    color: color.adjust($color-white, $lightness: -20%);
```
Comment (thibaudcolas) [inline client/scss/components/_header.scss:53]: This variable wasn’t used anywhere else anymore, and it’s a bit silly to tie together the colors of icons across different parts of the UI.

File: client/scss/core.scss [hunk 3]
```diff
@@ -140,6 +138,8 @@ These are classes for components.
 @import 'components/comments-notification-dropdown';
 @import 'components/bulk_actions';
 
+@import '../src/components/Sidebar/Sidebar';
```
Comment (thibaudcolas) [inline client/scss/core.scss:141]: I had to move this to the bottom because there are button styles that conflict with it.

File: client/scss/overrides/_utilities.legacy.scss [hunk 4]
```diff
@@ -1,3 +1,8 @@
+.u-hidden {
+  // stylelint-disable-next-line declaration-no-important
+  display: none !important;
+}
```
Comment (thibaudcolas) [inline client/scss/overrides/_utilities.legacy.scss:4]: All other `hidden` utilities have been switched to using Tailwind, as they were only used in very few spots after the legacy sidebar removal.

File: client/src/components/Draftail/Draftail.scss [hunk 5]
```diff
@@ -1,4 +1,4 @@
-$editor-z-index: $draftail-editor-z-index;
+$editor-z-index: 27;
```
Comment (thibaudcolas) [inline client/src/components/Draftail/Draftail.scss:1]: This used to be calculated based on another variable, which has been removed.

File: client/src/components/Sidebar/Sidebar.scss [hunk 6]
```diff
@@ -87,6 +87,14 @@
...
     display: none;
   }
+
+  .icon--menuitem {
+    width: 1.25em;
+    height: 1.25em;
+    min-width: 1.25em;
+    margin-inline-end: 0.5em;
+    vertical-align: text-top;
+  }
```
Comment (thibaudcolas) [inline client/src/components/Sidebar/Sidebar.scss:97]: This was also previously defined in `_main-nav.scss`, and duplicated in `MainMenu.scss` but only for the main menu icons.

File: wagtail/admin/static_src/wagtailadmin/scss/layouts/404.scss [hunk 7]
```diff
@@ -28,10 +28,15 @@
...
+.page404__logo {
   float: left;
   width: 400px;
   height: 500px;
+
+  // Media for Windows High Contrast mode
+  @media (forced-colors: $media-forced-colours) {
+    background-color: $system-color-link-text;
+  }
 }
```
Comment (thibaudcolas) [inline wagtail/admin/static_src/wagtailadmin/scss/layouts/404.scss:40]: This used to be defined in `_logo.scss`.

File: wagtail/admin/templates/wagtailadmin/404.html [hunk 8]
```diff
@@ -14,7 +14,9 @@
     <main class="page404__bg">
         <div class="w-w-full w-h-full w-max-w-6xl w-mx-auto w-flex w-items-center w-justify-center">
             {% block branding_logo %}
-                {% include "wagtailadmin/shared/animated_logo.html" %}
+                <div class="page404__logo w-hidden sm:w-block">
+                    <img src="{% versioned_static 'wagtailadmin/images/wagtail-logo.svg' %}" alt=""/>
+                </div>
```
Comment (thibaudcolas) [inline wagtail/admin/templates/wagtailadmin/404.html:19]: This `animated_logo.html` is no longer animated and is only used here.

File: wagtail/admin/ui/sidebar.py [hunk 9]
```diff
@@ -44,14 +44,22 @@ def js_args(self):
...
 class LinkMenuItem(MenuItem):
     def __init__(
-        self, name: str, label: str, url: str, icon_name: str = "", classnames: str = ""
+        self,
+        name: str,
+        label: str,
+        url: str,
+        icon_name: str = "",
+        classnames: str = "",
+        attrs: Mapping[str, Any] = None,
```
Comment (thibaudcolas) [inline wagtail/admin/ui/sidebar.py:53]: Not entirely sure `Mapping` is correct here – I don’t know how Telepath would serialise a `attrs` that’s not a dict.

File: wagtail/admin/wagtail_hooks.py [hunk 10]
```diff
@@ -917,7 +909,7 @@ def register_aging_pages_report_menu_item():
 
 @hooks.register("register_admin_menu_item")
 def register_reports_menu():
-    return ReportsMenuItem(_("Reports"), reports_menu, icon_name="site", order=9000)
+    return SubmenuMenuItem(_("Reports"), reports_menu, icon_name="site", order=9000)
```
Comment (thibaudcolas) [inline wagtail/admin/wagtail_hooks.py:912]: I’m uncertain of this. Did I misunderstand some interest of having a separate class for `ReportsMenuItem`?

**Comment by zerolab** [comment]
Screenshot 2022-04-08 at 09 28 09
😍

File: client/src/components/Draftail/Draftail.scss (hunk 5 above)
Comment (lb-) [reply client/src/components/Draftail/Draftail.scss:1]: Should we create new variable anyway?

**Comment by thibaudcolas** [comment]
Tracking remaining issues with the sidebar here: #8311.

**Comment by lb-** [comment]
Is it worth taking this opportunity to name this thing Menu instead of Sidebar?

Seeing as the python side is called Menu and most documentation refers to it as Menu or Main Menu.

File: wagtail/images/templates/wagtailimages/images/edit.html [hunk 11]
```diff
@@ -108,7 +108,7 @@ <h2 class="label no-float">{% trans "Focal point" %} <span class="w-font-normal"
             </div>
         </div>
 
-        <div class="row row-flush nice-padding u-hidden@sm">
+        <div class="row row-flush nice-padding sm:!w-hidden">
```
Comment (lb-) [inline wagtail/images/templates/wagtailimages/images/edit.html:111]: Nitpick but should these changes be in a separate PR?

File: wagtail/admin/templates/wagtailadmin/admin_base.html [hunk 12]
```diff
@@ -54,8 +52,8 @@
     <script src="{% versioned_static 'wagtailadmin/js/core.js' %}"></script>
     <script src="{% versioned_static 'wagtailadmin/js/vendor.js' %}"></script>
     <script src="{% versioned_static 'wagtailadmin/js/wagtailadmin.js' %}"></script>
-
-    {% main_nav_js %}
+    <script src="{% versioned_static 'wagtailadmin/js/telepath/telepath.js' %}"></script>
```
Comment (lb-) [inline wagtail/admin/templates/wagtailadmin/admin_base.html:55]: Should we bump telepath above wagtailadmin.js - it seems to be a pretty core import

File: wagtail/admin/templates/wagtailadmin/404.html (hunk 8 above)
Comment (lb-) [reply wagtail/admin/templates/wagtailadmin/404.html:19]: Maybe we should check any docs that reference overriding this template. Or add to upgrade considerations?

**Review by lb- — COMMENTED** [review]
Just some misc comments. 

Wow @thibaudcolas you are a machine.

File: wagtail/images/templates/wagtailimages/images/edit.html (hunk 11 above)
Comment (thibaudcolas) [reply wagtail/images/templates/wagtailimages/images/edit.html:111]: I thought this would be ok enough to refactor here as well considering those legacy utilities were only used in very few places aside from the sidebar.

File: wagtail/admin/templates/wagtailadmin/admin_base.html (hunk 12 above)
Comment (thibaudcolas) [reply wagtail/admin/templates/wagtailadmin/admin_base.html:55]: I tried to do this but got errors in the console – decided to keep the import order as-is for the time being. Ideally I think there should be fewer of those files to start with?

**Comment by thibaudcolas** [comment]
@lb- my understanding is that "sidebar" refers to the whole thing, while the menu / main menu is only a small part of it.

File: wagtail/admin/templates/wagtailadmin/404.html (hunk 8 above)
Comment (thibaudcolas) [reply wagtail/admin/templates/wagtailadmin/404.html:19]: The documented overrides are based on the `branding_logo` block so we should be good.

**Comment by lb-** [comment]
> @lb- my understanding is that "sidebar" refers to the whole thing, while the menu / main menu is only a small part of it.

Ok. That makes sense.

File: wagtail/admin/wagtail_hooks.py (hunk 10 above)
Comment (kaedroho) [reply wagtail/admin/wagtail_hooks.py:912]: Looks like this was introduced in https://github.com/wagtail/wagtail/commit/0db98757b41845d0f8ec631042f4cf0dc39186c3. It used to override the `is_shown` method but that override has since been removed.

I think this is fine to remove

File: client/src/index.ts [hunk 13]
```diff
@@ -10,6 +10,7 @@ import Portal from './components/Portal/Portal';
 import PublicationStatus from './components/PublicationStatus/PublicationStatus';
 import Transition from './components/Transition/Transition';
 import { initUpgradeNotification } from './components/UpgradeNotification';
+export { initSkipLink } from './includes/initSkipLink';
```
Comment (kaedroho) [inline client/src/index.ts:13]: For consistency, could we update all the other imports to re-export like this?

File: client/webpack.config.js [hunk 14]
```diff
@@ -187,14 +186,6 @@ module.exports = function exports(env, argv) {
     'panels',
     'streamfield.scss',
   );
-  sassEntry[getOutputPath('admin', 'css', 'sidebar')] = path.resolve(
-    'wagtail',
-    'admin',
-    'static_src',
-    'wagtailadmin',
-    'scss',
-    'sidebar.scss',
```
Comment (kaedroho) [inline client/webpack.config.js:196]: Should we delete `sidebar.scss` now?

**Review by kaedroho — APPROVED** [review]
Just a couple of minor comments but otherwise looks good to me!

File: wagtail/admin/wagtail_hooks.py (hunk 10 above)
Comment (thibaudcolas) [reply wagtail/admin/wagtail_hooks.py:912]: Thank you, should’ve known to check the blames

File: client/src/index.ts (hunk 13 above)
Comment (thibaudcolas) [reply client/src/index.ts:13]: Good point.

File: client/webpack.config.js (hunk 14 above)
Comment (thibaudcolas) [reply client/webpack.config.js:196]: Uh oh! Indeed.

**Comment by thibaudcolas** [comment]
Thank you both 🙌
